### PR TITLE
Pass the dtls record version to the record layer msg_callback function.

### PR DESCRIPTION
**Pull request by fwh-dc**

### CODE REVIEW

**Comment by fwh-dc** [comment]
Is dependent on #22238 for correct behaviour

**Comment by mattcaswell** [comment]
Hmm. Interesting.

This brings the behaviour into line with TLS which already does this. However the TLS behaviour seems to be new with 3.0 having been introduced via #12310. Prior to that DTLS and TLS just did the same thing which was to pass 0 for this version.

Interestingly it is explicitly [documented](https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set_msg_callback.html) that we will pass 0 for the version value when the "pseudo" content type `SSL3_RT_HEADER` is being used:

````
=item I<version>

The protocol version according to which the protocol message is
interpreted by the library such as B<TLS1_3_VERSION>, B<TLS1_2_VERSION> etc.
This is set to 0 for the SSL3_RT_HEADER pseudo content type (see NOTES below).
````

Note, that this version is a different version number to what is passed for non "pseudo" content types. In those cases we pass the actual protocol version number that has been negotiated - which could be different to the record layer protocol version number.

An extra consideration for DTLS is what we do for the equivalent call in `DTLSv1_listen()`:

https://github.com/openssl/openssl/blob/6bd07949e54f9958eb8a0f9a597ceb3910753ab0/ssl/d1_lib.c#L499-L501

Since TLS is already doing this and has done since 3.0, it seems reasonable to me that DTLS should be doing the same. In which case:

1) We should update the documentation to match what is actually being passed in the version field.
2) `DTLSv1_listen()` should do the same thing and pass the record layer version number in this call.

**Comment by fwh-dc** [comment]
Done. Please let me know if I got the terminology wrong.

**Review by mattcaswell — COMMENTED** [review]
This looks good.

I'm wondering whether or not we should treat this as a bug fix and backport to 3.1/3.0. The asymmetry between TLS/DTLS is kind of weird. But could anything be broken by this???

Also, maybe we should discuss the previous behaviour in the HISTORY section of the man page?

**Comment by fwh-dc** [comment]
IMO it is a logging interface so I wouldn't expect any functionality to be affected. But then again.. It wouldn't surprise me.

And yes I was aiming for consistency between TLS and DTLS.

Good suggestion to add to the HISTORY.

File: doc/man3/SSL_CTX_set_msg_callback.pod [hunk 1]
```diff
@@ -172,6 +173,9 @@ The pseudo content types B<SSL3_RT_QUIC_DATAGRAM>, B<SSL3_RT_QUIC_PACKET>,
 B<SSL3_RT_QUIC_FRAME_FULL>, B<SSL3_RT_QUIC_FRAME_HEADER> and
 B<SSL3_RT_QUIC_FRAME_PADDING> were added in OpenSSL 3.2.
 
+In versions previous to OpenSSL 3.3 I<cb> was called with 0 as I<version> for
```
Comment (t8m) [inline doc/man3/SSL_CTX_set_msg_callback.pod:176]: Hmm.... this could still make it to 3.2.

I would not change this in stable branches.

Comment (mattcaswell) [reply doc/man3/SSL_CTX_set_msg_callback.pod:176]: It's also not quite accurate, since for TLS it was changed in 3.0.

Comment (fwh-dc) [reply doc/man3/SSL_CTX_set_msg_callback.pod:176]: You are right. I will correct this.

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**